resource-managers/kubernetes/core/src/test/scala/org/apache/spark/deploy/kubernetes/submit/v2/ClientV2Suite.scala

```diff
@@ -0,0 +1,298 @@
+        (env.getName, env.getValue)
+      }.toMap
+      assert(envs(ENV_DRIVER_MEMORY) === s"${DRIVER_MEMORY_MB + DRIVER_MEMORY_OVERHEARD_MB}m")
+      assert(envs(ENV_DRIVER_MAIN_CLASS) === MAIN_CLASS)
+      assert(envs(ENV_DRIVER_ARGS) === APP_ARGS.mkString(" "))
+      assert(envs(ENV_SUBMIT_EXTRA_CLASSPATH) === DRIVER_EXTRA_CLASSPATH)
+    }
+  }
+
+  test("Created pod should apply custom annotations and labels") {
+    sparkConf.set(KUBERNETES_DRIVER_LABELS,
+      "label1=label1value,label2=label2value")
+    sparkConf.set(KUBERNETES_DRIVER_ANNOTATIONS,
+      "annotation1=annotation1value,annotation2=annotation2value")
+    val createdDriverPod = createAndGetDriverPod()
+    val labels = createdDriverPod.getMetadata.getLabels.asScala
+    assert(labels.size === 4)
+    // App ID is non-deterministic, but just check if it's set and is prefixed with the app name
+    val appIdLabel = labels(SPARK_APP_ID_LABEL)
+    assert(appIdLabel != null && appIdLabel.startsWith(APP_NAME) && appIdLabel != APP_NAME)
+    val appNameLabel = labels(SPARK_APP_NAME_LABEL)
+    assert(appNameLabel != null && appNameLabel == APP_NAME)
+    assert(labels("label1") === "label1value")
+    assert(labels("label2") === "label2value")
+    val annotations = createdDriverPod.getMetadata.getAnnotations.asScala
+    val expectedAnnotations = Map(
+      "annotation1" -> "annotation1value", "annotation2" -> "annotation2value")
+    assert(annotations === expectedAnnotations)
+  }
+
+  test("Driver JVM Options should be set in the environment.") {
+    sparkConf.set(org.apache.spark.internal.config.DRIVER_JAVA_OPTIONS, "-Dopt1=opt1value")
+    sparkConf.set("spark.logConf", "true")
+    val createdDriverPod = createAndGetDriverPod()
+    val maybeDriverContainer = getDriverContainer(createdDriverPod)
+    maybeDriverContainer.foreach { driverContainer =>
+      val maybeJvmOptionsEnv = driverContainer.getEnv
+        .asScala
+        .find(_.getName == ENV_DRIVER_JAVA_OPTS)
+      assert(maybeJvmOptionsEnv.isDefined)
+      maybeJvmOptionsEnv.foreach { jvmOptionsEnv =>
+        val jvmOptions = jvmOptionsEnv.getValue.split(" ")
+        jvmOptions.foreach { opt => assert(opt.startsWith("-D")) }
+        val optionKeyValues = jvmOptions.map { option =>
+          val withoutDashDPrefix = option.stripPrefix("-D")
+          val split = withoutDashDPrefix.split('=')
+          assert(split.length == 2)
+          (split(0), split(1))
+        }.toMap
+        assert(optionKeyValues("opt1") === "opt1value")
+        assert(optionKeyValues.contains("spark.app.id"))
+        assert(optionKeyValues("spark.jars") === MAIN_APP_RESOURCE)
+        assert(optionKeyValues(KUBERNETES_DRIVER_POD_NAME.key).startsWith(APP_NAME))
+        assert(optionKeyValues("spark.app.name") === APP_NAME)
+        assert(optionKeyValues("spark.logConf") === "true")
+      }
+    }
+  }
+
+  // Tests with local dependencies with the mounted dependency manager.
```

Comment:
There's still some more we can test here, particularly to check the driver container is picking up the resolved jars in the environment variables.